pkg/sql/catalog/descs/descriptor.go

```diff
@@ -86,12 +88,34 @@ func (tc *Collection) GetImmutableDescriptorByID(
 // The Required flag is ignored and always overridden.
 func (tc *Collection) getDescriptorsByID(
 	ctx context.Context, txn *kv.Txn, flags tree.CommonLookupFlags, ids ...descpb.ID,
-) (descs []catalog.Descriptor, err error) {
+) ([]catalog.Descriptor, error) {
+	descs := make([]catalog.Descriptor, len(ids))
+	if err := tc.getDescriptorsByIDWithSlice(ctx, txn, flags, descs, ids...); err != nil {
+		return nil, err
+	}
+	return descs, nil
+}
+
+// getDescriptorsByIDWithSlice is like getDescriptorsByID but takes the slice
+// into which the retreived descriptors will be stored. That slice must be the
+// same length as the ids. This allows callers seeking to get just one
+// descriptor to avoid an allocation by using a fixed-size array.
+func (tc *Collection) getDescriptorsByIDWithSlice(
+	ctx context.Context,
+	txn *kv.Txn,
+	flags tree.CommonLookupFlags,
+	descs []catalog.Descriptor,
+	ids ...descpb.ID,
+) (err error) {
```

Comment:
There's no point in maintaining both `getDescriptorsByID` and `getDescriptorsByIDWithSlice` now. Get rid of the former, take the latter and rename it to the former, and inline it everywhere. If need be define a helper `getDescriptorByID` which returns a `(desc, err)` pair.